src/overlay.c

```diff
@@ -0,0 +1,172 @@
+	struct wlr_scene_node *node = &seat->overlay.tree->node;
+	if (!wlr_renderer_is_pixman(server->renderer)) {
+		/* Hardware assisted rendering: Half transparent overlay */
+		wlr_scene_rect_set_size(seat->overlay.rect,
+			box->width, box->height);
+	} else {
+		/* Software rendering: Outlines */
+		multi_rect_set_size(seat->overlay.pixman_rect,
+			box->width, box->height);
+	}
+	if (node->parent != view->scene_tree->node.parent) {
+		wlr_scene_node_reparent(node, view->scene_tree->node.parent);
+		wlr_scene_node_place_below(node, &view->scene_tree->node);
+	}
+	wlr_scene_node_set_position(node, box->x, box->y);
+	wlr_scene_node_set_enabled(node, true);
+
+	cancel_pending_overlay(&seat->overlay);
+}
+
+static int
+handle_overlay_timeout(void *data)
+{
+	struct seat *seat = data;
+	show_overlay(seat, seat->overlay.view, &seat->overlay.box);
+	return 0;
+}
+
+static void
+show_overlay_delayed(struct seat *seat, struct view *view, struct wlr_box *box)
+{
+	if (seat->overlay.timer
+			&& seat->overlay.view == view
+			&& wlr_box_equal(box, &seat->overlay.box)) {
+		return;
+	}
+	cancel_pending_overlay(&seat->overlay);
+	seat->overlay.view = view;
+	seat->overlay.box = *box;
+	seat->overlay.timer = wl_event_loop_add_timer(
+		seat->server->wl_event_loop,
+		handle_overlay_timeout, seat);
+	wl_event_source_timer_update(seat->overlay.timer, 150);
+}
+
+void
+overlay_show(struct seat *seat, struct view *view)
+{
+	struct server *server = seat->server;
+
+	/*
+	 * TODO: cache return value of regions_from_cursor() or
+	 * edge_from_cursor() to eliminate duplicated calls to show_overlay() /
+	 * show_overlay_delayed().
+	 */
+	if (regions_should_snap(server)) {
+		/* Region overlay */
+		struct region *region = regions_from_cursor(server);
+		if (region) {
+			show_overlay(seat, view, &region->geo);
+			return;
+		}
+	} else {
+		/* Snap-to-edge overlay */
+		struct output *output;
+		enum view_edge edge = edge_from_cursor(seat, &output);
+		if (edge != VIEW_EDGE_INVALID) {
+			/* TODO: share logic with view_get_edge_snap_box() */
+			struct wlr_box box =
+				output_usable_area_in_layout_coords(output);
+			switch (edge) {
+			case VIEW_EDGE_RIGHT:
+				box.x += box.width / 2;
+				/* fallthrough */
+			case VIEW_EDGE_LEFT:
+				box.width /= 2;
+				break;
+			case VIEW_EDGE_DOWN:
+				box.y += box.height / 2;
+				/* fallthrough */
+			case VIEW_EDGE_UP:
+				box.height /= 2;
+				break;
+			case VIEW_EDGE_CENTER:
+				/* <topMaximize> */
+				break;
+			case VIEW_EDGE_INVALID:
+				/* not reached */
+				assert(false);
+			}
+			/*
+			 * Delay showing overlay to prevent flickering when
+			 * dragging view across output edges in multi-monitor
+			 * setup
+			 */
+			show_overlay_delayed(seat, view, &box);
+			return;
+		}
+	}
+	overlay_hide(seat);
+}
+
+void
+overlay_hide(struct seat *seat)
+{
+	cancel_pending_overlay(&seat->overlay);
+
+	struct server *server = seat->server;
+	struct wlr_scene_node *node = &seat->overlay.tree->node;
```

Comment:
This might segfault if we call `overlay_hide()` before `show_overlay()`, suggest simply returning if `!seat->overlay.tree`

Reply:
Done.
Actually, this doesn't segfault because node is never dereferenced, but it depended on the assumption of `&tree.node == &tree`.

Reply:
but `seat->overlay.tree` could be null theoretically, so any access to `->node` would already be trying to access a NULL pointer.

Reply:
I assumed `&(pointer)->(field)` is translated to `(pointer) + (offset of field)` so dereference doesn't happen at this point. I wonder if it is defined by C specification, though.

Reply:
That would actually make things worse, because if `pointer` were null and the compiler just blindly added an offset of a struct field, attempting to do would leave the value of `addr = &pointer->field` at a non-null that might circumvent later sanity checks on `addr` even though it is invalid. That would also be hard to debug.

I would hope that a compiler would actually try to deference a pointer when you use `->`, even if you are just computing an address of a field, to trigger a segfault where it would be expected rather than let a bogus address float around.